runtime/js-compute-runtime/builtins/shared/console.cpp

```diff
@@ -431,22 +434,390 @@ static bool console_out(JSContext *cx, unsigned argc, JS::Value *vp) {
+  } else {
+    label = "default";
+  }
+  if (!timer_map.contains(label)) {
+    std::string message = "No such label '";
+    message += label;
+    message += "' for console.timeLog()";
+    builtin_impl_console_log(Console::LogType::Warn, message.c_str());
+    args.rval().setUndefined();
+    return true;
+  }
+
+  // 2. Let startTime be timerTable[label].
+  auto startTime = timer_map[label];
+
+  // 3. Let duration be a string representing the difference between the current time and startTime,
+  // in an implementation-defined format.
+  auto duration = JS_Now() - startTime;
+  // 4. Let concat be the concatenation of label, U+003A (:), U+0020 SPACE, and duration.
+  std::string concat = label;
+  concat += ": ";
+  concat += std::to_string(duration);
+  concat += "ns";
+
+  std::string data = "";
+  if (args.length() > 1) {
+    auto length = args.length();
+    JS::RootedObjectVector visitedObjects(cx);
+    for (int i = 1; i < length; i++) {
+      JS::HandleValue arg = args.get(i);
+      std::string source = "";
+      auto result = ToSource(cx, source, arg, &visitedObjects);
+      if (result.isErr()) {
+        return false;
+      }
+      // strip quotes for direct string logs
+      if (source[0] == '"' && source[source.length() - 1] == '"') {
+        source = source.substr(1, source.length() - 2);
+      }
+      if (data.length()) {
+        data += " ";
+        data += source;
+      } else {
+        data += source;
+      }
+    }
+  }
+
+  // 5. Prepend concat to data.
+  concat += " ";
+  concat += data;
+
+  // 6. Perform Printer("timeLog", « concat »).
+  builtin_impl_console_log(Console::LogType::Log, concat.c_str());
+
+  args.rval().setUndefined();
+  return true;
+}
+
+static bool timeEnd(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS::CallArgs args = CallArgsFromVp(argc, vp);
+  // 1. Let timerTable be the associated timer table.
+  std::string label;
+  if (args.hasDefined(0)) {
+    auto label_val = args.get(0);
+    size_t length;
+    auto label_string = encode(cx, label_val, &length);
+    if (!label_string) {
+      return false;
+    }
+    label = label_string.get();
+  } else {
+    label = "default";
+  }
+  if (!timer_map.contains(label)) {
+    std::string message = "No such label '";
+    message += label;
+    message += "' for console.timeEnd()";
+    builtin_impl_console_log(Console::LogType::Warn, message.c_str());
+    args.rval().setUndefined();
+    return true;
+  }
+
+  // 2. Let startTime be timerTable[label].
+  // 3. Remove timerTable[label].
+  auto startTime = timer_map.extract(label).mapped();
+
+  // 4. Let duration be a string representing the difference between the current time and startTime,
+  // in an implementation-defined format.
+  auto duration = JS_Now() - startTime;
+  // 5. Let concat be the concatenation of label, U+003A (:), U+0020 SPACE, and duration.
+  std::string concat = label;
+  concat += ": ";
+  concat += std::to_string(duration);
+  concat += "ns";
```

Comment:
Typically this seems to work in seconds (node.js) or milliseconds (chrome / deno). With nanoseconds, this is 9 orders of magnitude smaller than milliseconds. Perhaps we should do some rounding or at least add a decimal place?

Reply:
node.js does dynamic formatting, it will use ms first and then if/when the period increases it will move to seconds/minutes/etc

Firefox always uses ms
Safari always uses fractional ms

I've updated our implementation to always use fractional ms 👍